### PR TITLE
feat(CapMan): `tenant_ids` in Snuba API requests

### PR DESCRIPTION
**Pull request by rahul-kumar-saini**
### Overview
- As part of the Capacity Management project, I've updated the Snuba SDK and updated the version in Snuba
- The `Request` object from Snuba SDK now expects (optionally for now) a new field `tenant_ids`
- https://github.com/getsentry/snuba-sdk/pull/112

### Changes
- Updated Snuba SDK dependency version to 1.0.5
- `AttributionInfo` now requires a new `tenant_ids` field
- Added a ddog metric to track how often we are *not* receiving the expected `tenant_ids`
    - This will help track down which features are not updated yet on Sentry

### CODE REVIEW

File: tests/datasets/entities/test_pluggable_entity.py [hunk 1]
```diff
@@ -124,6 +124,8 @@ def build_request(query_body: Mapping[str, str]) -> Request:
         query=query,
         snql_anonymized=snql_anonymized,
         query_settings=HTTPQuerySettings(referrer=""),
-        attribution_info=AttributionInfo(get_app_id("blah"), "blah", None, None, None),
+        attribution_info=AttributionInfo(
+            get_app_id("blah"), {"blah": "blah"}, "blah", None, None, None
```
Comment (volokluev) [inline tests/datasets/entities/test_pluggable_entity.py:128]: let's have these fields be a little more descriptive of what they're supposed to be mimicking. e.g. `{"tenant_type": "tenant_id"}:

**Review by volokluev — APPROVED** [review]

File: snuba/web/query.py [hunk 2]
```diff
@@ -201,13 +201,35 @@ def _run_query_pipeline(
...
+    metric will be removed once all API calls from Sentry do include this info.
+    """
+    if tenant_ids := request.attribution_info.tenant_ids:
+        if "referrer" not in tenant_ids or "organization_id" not in tenant_ids:
+            metrics.increment(
+                "request_without_tenant_ids",
+                tags={
+                    "referrer": request.referrer,
+                    "parent_api": request.attribution_info.parent_api or "None",
+                    "team": request.attribution_info.team or "None",
```
Comment (evanh) [inline snuba/web/query.py:226]: `team` and `feature` are unused/deprecated flags. No need to track them.

File: tests/request/test_schema.py [hunk 3]
```diff
@@ -12,10 +12,12 @@ def test_split_request() -> None:
         "team": "sns",
         "feature": "attribution",
         "app_id": "foobar",
+        "tenant_ids": {"tenant_type": "tenant_id"},
         "query": """MATCH (something) dontcare""",
     }
     schema = RequestSchema.build(HTTPQuerySettings)
     parts = schema.validate(payload)
+    print(parts.attribution_info)
```
Comment (evanh) [inline tests/request/test_schema.py:20]: 💀

File: snuba/web/query.py [hunk 4]
```diff
@@ -201,13 +201,35 @@ def _run_query_pipeline(
...
 
 
+def record_missing_tenant_ids(request: Request) -> None:
+    """
+    Used to track how often the new `tenant_ids` field is not included in
+    a Snuba Request. Ideally, all requests contain this information and this
+    metric will be removed once all API calls from Sentry do include this info.
+    """
+    if tenant_ids := request.attribution_info.tenant_ids:
+        if "referrer" not in tenant_ids or "organization_id" not in tenant_ids:
```
[the remark this answers is not in the document]
Comment (rahul-kumar-saini) [reply snuba/web/query.py:220]: What kind of queries will query multiple orgs? It's possible that those aren't the queries we really care about for capacity management anyway, I'm guessing it's internal stuff. We could assign a known default for a case like this to deal with separately in Snuba, what do you think @volokluev

[the remark this answers is not in the document]
Comment (volokluev) [reply snuba/web/query.py:220]: I would say let's ask in #discuss-backend what kind of queries query snuba across organizations. I think the known default makes sense as we would probably just have `referrer` be the tenant in that case but we should double check that our assumptions are correct and people are not making cross-org queries in the product

[the remark this answers is not in the document]
Comment (evanh) [reply snuba/web/query.py:220]: See this example of an entity that is used to query across orgs (currently used in release_health): https://github.com/getsentry/snuba/blob/master/snuba/datasets/configuration/metrics/entities/org_counters.yaml

Also this document on an upcoming feature that will do that: https://www.notion.so/sentry/Transaction-Name-Bias-44dcfe642a3848f493e054d8401d87cc

They are generally part of asynchronous jobs that run periodically, and definitely need to be included in capacity management to ensure they don't bog the cluster down whenever they run.

[the remark this answers is not in the document]
Comment (volokluev) [reply snuba/web/query.py:220]: Thanks for the example. Both of those look fine to me. 

The referrer for the cross-org query will effectively be the tenant in both of those cases. We can also create different allocation policies for different storages if there are storages that have different tenant definitions.

We can change how this metric is reported later.

**Pull request merged**